helix-core/src/main/java/org/apache/helix/controller/rebalancer/ConditionBasedRebalancer.java

```diff
@@ -0,0 +1,179 @@
+package org.apache.helix.controller.rebalancer;
+
+/*
+ * Licensed to the Apache Software Foundation (ASF) under one
+ * or more contributor license agreements.  See the NOTICE file
+ * distributed with this work for additional information
+ * regarding copyright ownership.  The ASF licenses this file
+ * to you under the Apache License, Version 2.0 (the
+ * "License"); you may not use this file except in compliance
+ * with the License.  You may obtain a copy of the License at
+ *
+ *     http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing,
+ * software distributed under the License is distributed on an
+ * "AS IS" BASIS, WITHOUT WARRANTIES OR CONDITIONS OF ANY
+ * KIND, either express or implied.  See the License for the
+ * specific language governing permissions and limitations
+ * under the License.
+ */
+
+import java.util.ArrayList;
+import java.util.Collections;
+import java.util.HashSet;
+import java.util.LinkedHashMap;
+import java.util.List;
+import java.util.Map;
+import java.util.Set;
+
+import org.apache.helix.HelixException;
+import org.apache.helix.controller.dataproviders.ResourceControllerDataProvider;
+import org.apache.helix.controller.rebalancer.condition.RebalanceCondition;
+import org.apache.helix.controller.stages.CurrentStateOutput;
+import org.apache.helix.model.IdealState;
+import org.apache.helix.model.LiveInstance;
+import org.apache.helix.model.Partition;
+import org.apache.helix.model.Resource;
+import org.apache.helix.model.ResourceAssignment;
+import org.apache.helix.model.StateModelDefinition;
+import org.apache.helix.zookeeper.datamodel.ZNRecord;
+import org.slf4j.Logger;
+import org.slf4j.LoggerFactory;
+
+public class ConditionBasedRebalancer extends AbstractRebalancer<ResourceControllerDataProvider> {
+  private static final Logger LOG = LoggerFactory.getLogger(ConditionBasedRebalancer.class);
+  private final RebalanceCondition _rebalanceCondition;
+
+  public ConditionBasedRebalancer(RebalanceCondition rebalanceCondition) {
```

Comment:
Let's make it generic? We can allow add a list of rebalance conditions.

Reply:
I guess this is already generic where we can specify simple or complex conditions. If you check the PR description, we can do the following:
Sample code to trigger the rebalance with condition `if (event type is CurrentStateChange OR CurrentStateChange)`
```
RebalanceCondition condition = new RebalanceConditionBuilder()
          .addCondition(() -> event.getEventType().equals(ClusterEventType.ClusterConfigChange))
          .addCondition(() -> event.getEventType().equals(ClusterEventType.CurrentStateChange))
          .buildOr();
Rebalancer rebalancer = new ConditionBasedRebalancer(condition);
```
We can also do nested condition as following:
```
BooleanSupplier condition1 = () -> 5 > 3;
BooleanSupplier condition2 = () -> 2 > 3;
BooleanSupplier condition3 = () -> event.getEventType().equals(ClusterEventType.ClusterConfigChange);
BooleanSupplier condition4 = () -> event.getEventType().equals(ClusterEventType.CurrentStateChange);

RebalanceCondition andCondition = new RebalanceConditionBuilder()
        .addCondition(condition1)
        .addCondition(condition3)
        .buildAnd();

RebalanceCondition orCondition = new RebalanceConditionBuilder()
        .addCondition(condition2)
        .addCondition(condition4)
        .buildOr();

RebalanceCondition nestedCondition = new RebalanceConditionBuilder()
        .addAndCondition(andCondition, orCondition)
        .buildAnd();
Rebalancer rebalancer = new ConditionBasedRebalancer(nestedCondition);
```
Let me know how do you think, open to any suggestions

Reply:
updated